Fixing error message verbiage

krmahadevan opened the pull request:
Fixes # .
### Did you remember to?
- [ ] Add test case(s)
- [X] Update `CHANGES.txt`

Fixes #1165

krmahadevan commented:
@cbeust / @juherr - can you please help take a look at this ?

juherr commented:
I relaunched travis. Tests are sometimes failing.

About the pull request, I think you can add an integration test with exception checking (maybe in test/inject package).

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -191,12 +192,12 @@ private static void checkParameterTypes(String methodName,
-          + " were supplied in the "
-          + methodAnnotation
-          + " annotation.");
+      throw new TestNGException(
```
I'm not sure this exception is related to "if (parameterNames.length != totalLength)".
I think we can have more context somewhere.

krmahadevan replied:
I fixed this.

krmahadevan commented:
@juherr - I have addressed your review comments. Can you please take a look ?

krmahadevan commented:
@juherr - gentle reminder :)

juherr commented:
Sorry for the delay.

Your change is a bit better but I think we can do better: as we know the type, we can manage some of them.
For example for #1165, we can explain `ITestNGMethod` is not supported on `@AfterX` methods (maybe some other types have the same behavior).
Another example with #1181 where @nitinverma proposed to add a warning.
I feel both logics should be located at the same place.

I hope I'm clear enough ^_^

krmahadevan commented:
@juherr 
No its not clear on the expectations of this issue.
W.r.t explaining the type of parameters that can be injected for a configuration method are saying that the error message basically captures all the combinations that are captured in the [documentation](http://testng.org/doc/documentation-main.html#native-dependency-injection) ? [ quoting the documentation for reference ]

> 1. Any @Before method or @Test method can declare a parameter of type ITestContext.
> 2. Any @AfterMethod method can declare a parameter of type ITestResult, which will reflect the result of the test method that was just run.
> 3. Any @Before and @After methods can declare a parameter of type XmlTest, which contain the current <test> tag.
> 4. Any @BeforeMethod (and @AfterMethod) can declare a parameter of type java.lang.reflect.Method. This parameter will receive the test method that will be called once this @BeforeMethod finishes (or after the method as run for @AfterMethod).
> 5. Any @BeforeMethod can declare a parameter of type Object[]. This parameter will receive the list of parameters that are about to be fed to the upcoming test method, which could be either injected by TestNG, such as java.lang.reflect.Method or come from a @DataProvider.
> 6. Any @DataProvider can declare a parameter of type ITestContext or java.lang.reflect.Method. The latter parameter will receive the test method that is about to be invoked. 

Wouldn't that be kind of redundancy ? Especially when the error message does point back to a link to the documentation which a user can refer to ?

W.r.t what @nitinverma commented [here](https://github.com/cbeust/testng/issues/1181#issuecomment-253276860) I am guessing that it might not be applicable here, because I think the method `org.testng.internal.Parameters#checkParameterTypes` wherein I am making changes as part of this PR, is not called for a data driven method but only called for a method that is either annotated with a config annotation or one that involves parameters. Atleast that's my limited understanding.

So please clarify so that I can wrap work on this PR

juherr commented:
Sorry. I try again :)

The goal is to have a better feedback to the users on injection (native, parameters and/or data provider) of method params.
Currently, if I'm right, we have `org.testng.internal.Parameters#checkParameterTypes` in one side and `org.testng.internal.reflect.MethodMatcher` in the other side. From my point of view, parameters injection should just be an internal data provider (but I didn't try any change in this direction for the moment).
A first step could be on params management.

As example (good idea to quote the doc!):

> 1. Any @Before method or @Test method can declare a parameter of type ITestContext.

To check, but I think it is possible to inject `ITestContext`.
If it is possible -> fix the doc
If not -> An exception is expected and should contain message like "Not possible to inject ITestContext on `@AfterX` methods"

> 2. Any @AfterMethod method can declare a parameter of type ITestResult, which will reflect the result of the test method that was just run.

"Not possible to inject ITestContext, except on `@AfterMethod` methods"

> 4. Any @BeforeMethod (and @AfterMethod) can declare a parameter of type java.lang.reflect.Method. This parameter will receive the test method that will be called once this @BeforeMethod finishes (or after the method as run for @AfterMethod).

"Not possible to inject java.lang.reflect.Method, except on `@BeforeMethod` and `@AfterMethod` methods"

...

One idea could be to have an injection whitelist for each configuration methods and a specific message for each supported injection types (ITestContext, ITestResult, ...).

Maybe param injection on configuration methods can be distinct from param injection on test methods. It is not clear enough in my head for the moment :)

krmahadevan commented:
Julien,
I will work on this and update this PR. I may probably send a different PR
to add more details to our native-injection related documentation as well [
Am thinking of translating this into a tabular format so that its a bit
more easier to understand what can be injected into what method ].


Thanks & Regards
Krishnan Mahadevan

"All the desirable things in life are either illegal, expensive, fattening
or in love with someone else!"
My Scribblings @ http://wakened-cognition.blogspot.com/
My Technical Scribbings @ http://rationaleemotions.wordpress.com/

On Sat, Oct 15, 2016 at 1:36 PM, Julien Herr <notifications@github.com>
wrote:

> Sorry. I try again :)
>
> The goal is to have a better feedback to the users on injection (native,
> parameters and/or data provider) of method params.
> Currently, if I'm right, we have org.testng.internal.
> Parameters#checkParameterTypes in one side and
> org.testng.internal.reflect.MethodMatcher in the other side. From my
> point of view, parameters injection should just be an internal data
> provider (but I didn't try any change in this direction for the moment).
> A first step could be on params management.
>
> As example (good idea to quote the doc!):
>
>
>    1. Any @Before <https://github.com/Before> method or @Test
>    <https://github.com/Test> method can declare a parameter of type
>    ITestContext.
>
> To check, but I think it is possible to inject ITestContext.
> If it is possible -> fix the doc
> If not -> An exception is expected and should contain message like "Not
> possible to inject ITestContext on @AfterX methods"
>
>
>    1. Any @AfterMethod method can declare a parameter of type
>    ITestResult, which will reflect the result of the test method that was just
>    run.
>
> "Not possible to inject ITestContext, except on @AfterMethod methods"
>
>
>    1. Any @BeforeMethod (and @AfterMethod) can declare a parameter of
>    type java.lang.reflect.Method. This parameter will receive the test method
>    that will be called once this @BeforeMethod finishes (or after the method
>    as run for @AfterMethod).
>
> "Not possible to inject java.lang.reflect.Method, except on @BeforeMethod
> and @AfterMethod methods"
>
> ...
>
> One idea could be to have an injection whitelist for each configuration
> methods and a specific message for each supported injection types
> (ITestContext, ITestResult, ...).
>
> Maybe param injection on configuration methods can be distinct from param
> injection on test methods. It is not clear enough in my head for the moment
> :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cbeust/testng/pull/1199#issuecomment-253969844>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA3hzQYnwzGhH7z4JSld7yscdBytHRMeks5q0Il8gaJpZM4KTRhL>
> .
>

krmahadevan commented:
@juherr -sorry on the delay on this. Kind of missed this PR. Please check if the new changes are fine.

juherr commented:
@krmahadevan No prob and the need is not so clear (my fault!). And the issue is not so important :)

BTW, could you have a look on my latest open issues? I'd like to have your opinion on them.

krmahadevan commented:
@juherr - I have added my opinion to the issue #1223 If there is some other issue that you had in mind please let me know. I will try my best to add whatever I can.

krmahadevan commented:
@juherr - Can you please let me know if this PR can be merged ? [ since I have anyways put in effort to fix this issue, I was hoping that we might as well just get this merged as well ]

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -43,6 +38,42 @@
+      AfterMethod.class
+  };
+
+  private static Map<String, List<Class>> mapping = new HashMap<>();
```
`private static Map<String, List<Class<?>>> mapping = Maps.new...();`
or better, if it works:
`private static Map<String, List<Class<? extends Annotation>>> mapping = Maps.new...();`

krmahadevan replied:
I have changed this to 

`private static Map<String, List<Class>> mapping = Maps.newHashMap();`

The other option will not work because the keys are annotations and Not the values in the map.

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -43,6 +38,42 @@
+  private static Map<String, List<Class>> mapping = new HashMap<>();
+
+  static {
+    List<Class> ctxTest = Arrays.asList(new Class[] {ITestContext.class, XmlTest.class});
```
```java
List<Class<?>> ctxTest = Arrays.asList(ITestContext.class, XmlTest.class);
List<Class<?>> mParams = Arrays.asList(ITestContext.class, XmlTest.class, Method.class, Object[].class);
```

krmahadevan replied:
This cannot be done, because I am being thrown errors by IntelliJ wherein for the first statement 
`List<Class<?>> ctxTest = Arrays.asList(ITestContext.class, XmlTest.class);` it says that expected was `List<?>` but actual values were `List<? extends Serializable>`
and for the second one, there's a Object[] array. Its difficult to maintain parity across both.

So I chose to leave this as is and Not address this comment.

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -190,6 +232,21 @@ private static void checkParameterTypes(String methodName,
+      //apply our logic only when @Parameters annotation is not involved.
+      String errPrefix = "Cannot inject " + methodAnnotation + " annotated Method [" + methodName + "] with "
+            + Arrays.toString(parameterTypes);
+      if (mapping.containsKey(methodAnnotation)) {
```
Replace `String methodAnnotation` by `Annotation methodAnnotation`, it will be stronger.

krmahadevan replied:
Tried addressing this and its triggering an avalanche of changes (including changing some public APIs). So I would prefer not to extend the scope of this PR beyond adding a better error message.

juherr replied:
Ok!

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -200,6 +257,21 @@ private static void checkParameterTypes(String methodName,
     }
   }
 
+  private static String prettyFormat(List<Class> classes) {
```
`List<Class<?>>`

krmahadevan replied:
Not able to incorporate this because I didn't change `mapping` value type to `<Class<?>>`

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -200,6 +257,21 @@ private static void checkParameterTypes(String methodName,
 
+  private static String prettyFormat(List<Class> classes) {
+    StringBuilder builder = new StringBuilder("<");
+    if (classes.size() == 1) {
```
No need to check against size=1, the loop will work if the list size = 1 too.
But currently, the method will fail if the list is empty ;)

krmahadevan replied:
The explicit check for `size ==1` was done because if there was just one element in the array then we need a pretty formatted output like `<Foo>` (without the `,`)
Yes the method will fail if the list is empty, but since the list is currently controlled via static data, I thought it would be ok.

juherr replied:
Yes, but in case of `size == 1`, the loop has no element and only  `builder.append(classes.get(length-1).getSimpleName());` will be triggered, right?

At least, add `.getSimpleName()` in case of 1 and maybe return an empty string in case of 0.

juherr commented on src/main/java/org/testng/internal/Parameters.java:
```diff
@@ -200,6 +257,21 @@ private static void checkParameterTypes(String methodName,
+    } else {
+      int length = classes.size();
+      for (int i=0; i < length - 1; i++) {
+        builder.append(classes.get(i).getSimpleName()).append(",");
```
I prefer ", " (with a white space)

krmahadevan replied:
Took care of this.

krmahadevan commented:
@juherr - i have addressed some of the review comments and for the ones wherein I didnt fix the review comment I have added my comments. Please let me know if there is anything additional required.

juherr commented:
10 tests are failing on travis with this change :(

krmahadevan commented:
Yikes.. my worst fears came true :( I will look at the reasons for the
failures and update this PR Julien.

​~Krishnan​


On Mon, Nov 7, 2016 at 2:17 PM, Julien Herr <notifications@github.com>
wrote:

> 10 tests are failing on travis with this change :(
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cbeust/testng/pull/1199#issuecomment-258777359>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA3hzSJMleh8l107zWUOdSKIcbWPPvq6ks5q7uWtgaJpZM4KTRhL>
> .
>

krmahadevan commented:
@juherr - Please give me some more time. I will get around to investigating why the tests are failing and update you accordingly. For now I have just made sure that there are no conflicts on this PR.

krmahadevan commented:
@juherr - I fixed the test failures and ran the tests locally just to be absolutely sure. Please help get this merged.

juherr commented:
@krmahadevan Where did you fix them? The latest commit is 4days old and is only working on travis with jdk8.

krmahadevan commented:
@juherr - I ended up amending the previous commit and pushed the changes forcibly, because it was only a spacing issue in the expected error message. :(
Should I just close this Pull request and open a new one after squashing all the commits? If needed I can call out link to this PR in the new one so that the comment history can be referred if needed. Please let me know.

juherr commented:
It's ok. I didn't catch it. Just fix test with jdk7

krmahadevan commented:
@juherr  -  I am not able to find out what failed..
I took a look at the details here : https://travis-ci.org/cbeust/testng/jobs/175058671

and here's what I see
```
1357 tests completed, 1 failed
Finished generating test XML results (0.135 secs) into: /home/travis/build/cbeust/testng/build/test-results/test
Generating HTML test report...
Finished generating test html results (0.618 secs) into: /home/travis/build/cbeust/testng/build/reports/tests/test
:test FAILED
:test (Thread[Daemon worker,5,main]) completed. Took 2 mins 46.176 secs.
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/travis/build/cbeust/testng/build/reports/tests/test/index.html
```

Do you know how I can get access to seeing this file from the browser ?
```
file:///home/travis/build/cbeust/testng/build/reports/tests/test/index.html
```
Atleast that would help me figure out what is failing. 

In Jenkins I would usually go to the workspace and find files, but I am not able to figure out how to do the same in Travis. It would be great if you could please help me.

There is just one test that is failing and its the same result locally as well. The test that is failing is 
```
DataProviderTest. shouldNotThrowConcurrentModificationException

```

juherr commented:
> Do you know how I can get access to seeing this file from the browser ?

It is not possible. That's the problem with travis.

> DataProviderTest. shouldNotThrowConcurrentModificationException

Again this one :/ I am relaunching travis and cross my fingers.

krmahadevan commented:
It failed again.. and it looks like its the same failure @juherr

krmahadevan commented:
@juherr - I have finally gotten a green build for this delivery.
@cbeust - Julien has approved this PR. Can you please help me with getting this merged as well ?

juherr commented:
LGTM :+1:  Ping @cbeust
